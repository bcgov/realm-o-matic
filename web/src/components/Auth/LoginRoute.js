// eslint-disable-next-line
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import implicitAuthManager from '../../utils/auth';

export const LoginRoute = ({ match }) => {
  const idp = match.params.idp;
  if (idp) implicitAuthManager.config.kcIDPHint = idp;
  window.location.assign(implicitAuthManager.getSSOLoginURI());
  return <div>Your are being redirected to KeyCloak login page...</div>;
};

LoginRoute.propTypes = {
  match: PropTypes.object.isRequired,
};
