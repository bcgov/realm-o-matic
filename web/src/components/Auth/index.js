export { default as AuthModal } from './AuthModal';
export { LoginRoute } from './LoginRoute';
export { default as ProtectedRoute } from './ProtectedRoute';
