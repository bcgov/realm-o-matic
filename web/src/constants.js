//
// Realm-o-Matic
//
// Copyright Â© 2019 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Shelly Xue Han
//

import config from './config.json';

export const SSO_CONFIG = {
  baseURL: process.env.REACT_APP_SSO_BASE_URL || config.ssoBaseUrl,
  realmName: process.env.REACT_APP_SSO_REALM_NAME || config.ssoRealmName,
  clientId: process.env.REACT_APP_SSO_CLIENT_ID || config.ssoClientId,
};

export const API = {
  BASE_URL: () =>
    process.env.NODE_ENV === 'development'
      ? 'http://localhost:8000/api/v1/'
      : `${window.location.origin}/api/v1/`,
  TIME_OUT: 40000,
  IDP: () => 'idps',
};

export const TEST_IDS = {
  APP: {
    LOGIG: 'login-button',
    GET_IDPS: 'get-idps-button',
  },
};

export const APP_INFO = {
  NAME: 'Realm-o-Matic',
};

export const SSO_IDP = {
  GITHUB: 'github',
  IDIR: 'idir',
};

export const FOOTER_LINKS = [
  {
    name: 'Home',
    link: '.',
  },
  {
    name: 'Disclaimer',
    link: 'https://www2.gov.bc.ca/gov/content/home/disclaimer',
  },
  {
    name: 'Privacy',
    link: 'https://www2.gov.bc.ca/gov/content/home/privacy',
  },
  {
    name: 'Accessibility',
    link: 'https://www2.gov.bc.ca/gov/content/home/accessibility',
  },
  {
    name: 'Copyright',
    link: 'https://www2.gov.bc.ca/gov/content/home/copyright',
  },
  {
    name: 'Contact Us',
    link: 'https://github.com/bcgov/realm-o-matic',
  },
];
