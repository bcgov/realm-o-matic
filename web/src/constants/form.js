//
// Realm-o-Matic
//
// Copyright Â© 2019 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Shelly Xue Han
//

const basicInfo = {
  name: 'realmInfo',
  title: 'Realm Information',
  elements: [
    {
      type: 'text',
      name: 'realmId',
      title: 'The Unique ID of your realm',
      isRequired: true,
      enableIf: 'false',
    },
    {
      type: 'text',
      name: 'displayName',
      title: 'The realm display name',
      isRequired: true,
    },
    {
      type: 'text',
      name: 'adminUser',
      title: 'The administrator of the realm (IDIR username)',
      isRequired: true,
    },
    {
      type: 'checkbox',
      name: 'idps',
      title: 'Required identity providers',
      isRequired: true,
      choices: ['IDIR', 'GitHub', 'BCeID'],
    },
    {
      type: 'text',
      name: 'po',
      title: 'The product owner of project',
      isRequired: true,
      validators: [
        {
          type: 'email',
        },
      ],
    },
  ],
};

const requesterInfo = {
  name: 'requesterInfo',
  title: 'Requester Information',
  elements: [
    {
      type: 'panel',
      name: 'requester',
      elements: [
        {
          type: 'text',
          name: 'requesterEmail',
          title: 'Email',
          isRequired: true,
          enableIf: 'false',
          colCount: 0,
          validators: [
            {
              type: 'email',
            },
          ],
        },
        {
          type: 'text',
          name: 'requesterIDIR',
          title: 'IDIR Username',
          // startWithNewLine: false,
          isRequired: true,
          enableIf: 'false',
          colCount: 0,
        },
        {
          type: 'text',
          name: 'requesterLastName',
          title: 'Last Name (optional)',
          colCount: 0,
        },
        {
          type: 'text',
          name: 'requesterFirstName',
          title: 'First Name (optional)',
          colCount: 0,
        },
      ],
      title: 'Requester Information',
    },
  ],
};

export const formJson = {
  showProgressBar: 'top',
  showQuestionNumbers: 'off',
  pages: [basicInfo, requesterInfo],
};

export const formDataToRequest = {
  id: 'id',
  realm: {
    id: 'realmId',
    displayName: 'displayName',
    adminUser: 'adminUser',
    idps: 'idps',
    po: 'po',
  },
  requester: {
    username: 'requesterIDIR',
    email: 'requesterEmail',
    firstName: 'requesterFirstName',
    lastName: 'requesterLastName',
  },
};

export const requestToFormData = {
  id: 'id',
  realmId: 'realm.id',
  displayName: 'realm.displayName',
  requesterEmail: 'requester.email',
  requesterIDIR: 'requester.username',
  requesterLastName: 'requester.lastName',
  requesterFirstName: 'requester.firstName',
  adminUser: 'realm.adminUser',
  idps: 'realm.adminUser',
  po: 'realm.po',
};
