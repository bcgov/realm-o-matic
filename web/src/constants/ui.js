//
// Realm-o-Matic
//
// Copyright Â© 2019 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Shelly Xue Han
//

export const TEST_IDS = {
  AUTH: {
    MODAL: 'login-modal',
    IDIR_LOGIN: 'idir-login-button',
    GITHUB_LOGIN: 'github-login-button',
  },
  APP: {
    NAME: 'app-name',
    LOGO: 'gov-logo',
    ACCOUNT: 'account-dropdown',
    NEW_REQUEST: 'new-request-button',
    HIDDEN_NEW_REQUEST: 'hidden-new-request-button',
    LOADER: 'load-indicator',
    EMPTY: 'empty-content',
    POPUP: 'status-popup',
  },
  REQUEST: {
    MESSAGE: 'request-message',
    FORM: 'request-form',
    FORM_LIST: 'request-list',
    APPROVAL: 'request-approval',
    REJECT: 'request-rejection',
    CANCEL: 'go-back',
  },
};

export const APP_INFO = {
  NAME: 'Realm-o-Matic',
  CONTACT_LINK: 'https://chat.pathfinder.gov.bc.ca',
};

export const FOOTER_LINKS = [
  {
    name: 'Home',
    link: '.',
  },
  {
    name: 'Disclaimer',
    link: 'https://www2.gov.bc.ca/gov/content/home/disclaimer',
  },
  {
    name: 'Privacy',
    link: 'https://www2.gov.bc.ca/gov/content/home/privacy',
  },
  {
    name: 'Accessibility',
    link: 'https://www2.gov.bc.ca/gov/content/home/accessibility',
  },
  {
    name: 'Copyright',
    link: 'https://www2.gov.bc.ca/gov/content/home/copyright',
  },
  {
    name: 'Contact Us',
    link: 'https://github.com/bcgov/realm-o-matic',
  },
];
