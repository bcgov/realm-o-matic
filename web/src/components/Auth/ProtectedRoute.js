import React from 'react';
import PropTypes from 'prop-types';
import { Route, Redirect } from 'react-router-dom';
import { ACCESS_CONTROL } from '../../constants/auth';

/**
 * Protected Route, redirect back to root page is not authorized for the content
 * https://medium.com/@leonardobrunolima/react-tips-how-to-protect-routes-for-unauthorized-access-with-react-router-v4-73c0d451e0a2
 * @param {Component} component the Component to render
 * @param {String} authCode the authorization code
 * @param {Object} rest the other objects passed in, e.g. the path of route
 */
export const ProtectedRoute = ({ component: Component, authCode, ...rest }) => {
  const isRoled =
    authCode === ACCESS_CONTROL.REQUESTER_ROLE || authCode === ACCESS_CONTROL.REVIEWER_ROLE;
  return (
    <Route {...rest} render={props => (isRoled ? <Component {...props} /> : <Redirect to="/" />)} />
  );
};

ProtectedRoute.propTypes = {
  authCode: PropTypes.string,
};

ProtectedRoute.defaultProps = {
  authCode: null,
};

export default ProtectedRoute;
