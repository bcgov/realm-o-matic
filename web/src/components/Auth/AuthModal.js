import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { Link } from 'react-router-dom';
import { Modal, Button } from 'semantic-ui-react';
import { SSO_IDP } from '../../constants/auth';
import { TEST_IDS } from '../../constants/ui';

const StyledModal = styled(Modal)`
  &&& {
    width: 400px;
    text-align: center;
    padding: 10px 0;
  }
  .content {
    margin: 30px 0;
  }
  .ui.button {
    color: white;
    background-color: #036;
  }
`;

export const AuthModal = ({ isAuthenticated }) => {
  const buttons = (
    <Button.Group onActionClick={null}>
      <Link to={{ pathname: `/login/${SSO_IDP.GITHUB}` }}>
        <Button data-testid={TEST_IDS.AUTH.GITHUB_LOGIN}>GitHub</Button>
      </Link>
      <Button.Or />
      <Link to={{ pathname: `/login/${SSO_IDP.IDIR}` }}>
        <Button data-testid={TEST_IDS.AUTH.IDIR_LOGIN}>IDIR</Button>
      </Link>
    </Button.Group>
  );

  const options = {
    open: !isAuthenticated,
    closeOnEscape: false,
    closeOnDimmerClick: false,
    size: 'tiny',
    dimmer: 'blurring',
    header: 'Welcome to Realm-o-Matic',
    content: 'Please login with one of the following options before proceeding to the app.',
    actions: buttons,
  };
  return <StyledModal data-testid={TEST_IDS.AUTH.MODAL} {...options} />;
};

AuthModal.propTypes = {
  isAuthenticated: PropTypes.bool.isRequired,
};

export default AuthModal;
