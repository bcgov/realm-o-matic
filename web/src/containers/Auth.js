import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { ACCESS_CONTROL } from '../constants/auth';
import { SpinLoader } from '../components/UI';
import { AuthModal } from '../components/Auth/AuthModal';
import { authorizationAction } from '../actionCreators';

export class Auth extends Component {
  static displayName = '[Component Auth]';
  componentDidMount = () => {
    const {
      isAuthenticated,
      userInfo,
      userId,
      authorizationStarted,
      authCode,
      errorMessage,
      authorizationAction,
    } = this.props;
    if (isAuthenticated && !authorizationStarted && !authCode && !errorMessage) {
      authorizationAction(userId, userInfo.roles);
    }
  };
  render() {
    const { isAuthenticated, authorizationStarted, authCode, errorMessage } = this.props;

    let redirect = null;
    if (authCode === ACCESS_CONTROL.NO_ROLE) redirect = <Redirect to="/notAuthorized" />;
    else if (
      authCode === ACCESS_CONTROL.REQUESTER_ROLE ||
      authCode === ACCESS_CONTROL.REVIEWER_ROLE
    ) {
      redirect = <Redirect to="/home" />;
    }

    let content;
    if (!isAuthenticated) {
      content = <AuthModal isAuthenticated={isAuthenticated} />;
    } else {
      if (authorizationStarted) content = <SpinLoader text="Loading your user info..." />;
      else if (errorMessage) content = <p>{errorMessage}</p>;
      else content = redirect;
    }
    return <div>{content}</div>;
  }
}

const mapStateToProps = state => {
  return {
    // Authentication:
    isAuthenticated: state.authentication.isAuthenticated,
    userInfo: state.authentication.userInfo,
    userId: state.authentication.userId,
    // Authorization:
    authorizationStarted: state.authorization.authorizationStarted,
    authCode: state.authorization.authCode,
    errorMessage: state.authorization.errorMessage,
  };
};

const mapDispatchToProps = dispatch => {
  return bindActionCreators(
    {
      authorizationAction,
    },
    dispatch
  );
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Auth);
