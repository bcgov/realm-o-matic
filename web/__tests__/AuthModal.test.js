import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import { BrowserRouter } from 'react-router-dom';
// eslint-disable-next-line
import Portal from 'semantic-ui-react/dist/commonjs/addons/Portal/Portal';
import { AuthModal } from '../src/components/Auth/AuthModal';
import { TEST_IDS } from '../src/constants/ui';
import { KC_IDP_NAMES } from '../src/constants/form';

jest.mock('semantic-ui-react/dist/commonjs/addons/Portal/Portal', () => ({ children }) => children);

describe('AuthModal Component', () => {
  it('renders without crashing', () => {
    const { container, getByText, getByTestId } = render(
      <BrowserRouter>
        <AuthModal isAuthenticated={false} />
      </BrowserRouter>
    );

    getByText('Welcome to Realm-o-Matic');
    expect(getByTestId(TEST_IDS.AUTH.GITHUB_LOGIN)).toHaveTextContent(KC_IDP_NAMES.GITHUB);
    expect(getByTestId(TEST_IDS.AUTH.IDIR_LOGIN)).toHaveTextContent(KC_IDP_NAMES.IDIR);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('redirect when clicking on the idp buttons', () => {
    const { getByTestId } = render(
      <BrowserRouter>
        <AuthModal isAuthenticated={false} />
      </BrowserRouter>
    );
    const githubButton = getByTestId(TEST_IDS.AUTH.GITHUB_LOGIN);

    fireEvent.click(githubButton);
    expect(window.location.href).toContain('/login/github');
  });
});
