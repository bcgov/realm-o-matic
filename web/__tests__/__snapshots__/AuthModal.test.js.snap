// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthModal Component renders without crashing 1`] = `
<div
  class="ui page modals dimmer transition visible active"
>
  <div
    class="ui tiny modal transition visible active"
    data-testid="login-modal"
  >
    <div
      class="header"
    >
      Login first
    </div>
    <div
      class="content"
    >
      <p>
        Please Login Before Proceeding
      </p>
    </div>
    <div
      class="actions"
    >
      <a
        href="/login/github"
      >
        <button
          class="ui button"
          data-testid="github-login-button"
        >
          GitHub
        </button>
      </a>
      <a
        href="/login/idir"
      >
        <button
          class="ui button"
          data-testid="idir-login-button"
        >
          IDIR
        </button>
      </a>
    </div>
  </div>
</div>
`;
